clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -388,43 +391,39 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         // by markers.
         bool have_cond_branch_target = false;
         addr_t cond_branch_target = 0;
-        if (shard->prev_instr_.instr.addr != 0 /*first*/ &&
-            type_is_instr_direct_branch(shard->prev_instr_.instr.type) && has_encoding) {
-            if (shard->prev_instr_.instr.encoding_is_new)
+        if (type_is_instr_direct_branch(memref.instr.type) && expect_encoding) {
+            // We do not bother to support legacy traces without encodings.
+            if (memref.instr.encoding_is_new)
                 shard->branch_target_cache.erase(trace_pc);
             auto cached = shard->branch_target_cache.find(trace_pc);
             if (cached != shard->branch_target_cache.end()) {
                 have_cond_branch_target = true;
                 cond_branch_target = cached->second;
             } else {
-                if ((shard->prev_instr_decoded_ == nullptr ||
-                     !opnd_is_pc(instr_get_target(shard->prev_instr_decoded_)))) {
-
+                if (cur_instr_decoded == nullptr ||
+                    !opnd_is_pc(instr_get_target(cur_instr_decoded.get()))) {
                     // Neither condition should happen but they could on an invalid
                     // encoding from raw2trace or the reader so we report an
                     // invariant rather than asserting.
                     report_if_false(shard, false, "Branch target is not decodeable");
                 } else {
                     have_cond_branch_target = true;
                     cond_branch_target = reinterpret_cast<addr_t>(
-                        opnd_get_pc(instr_get_target(shard->prev_instr_decoded_)));
+                        opnd_get_pc(instr_get_target(cur_instr_decoded.get())));
                     shard->branch_target_cache[trace_pc] = cond_branch_target;
                 }
             }
         }
         bool saw_repeated_syscall_instrs_with_same_pc = false;
-        if (shard->prev_instr_.instr.addr != 0 /*first*/) {
-            if (next_pc != nullptr) {
-                if (instr_is_syscall(&cur_instr_decoded) &&
-                    memref.instr.addr == shard->prev_instr_.instr.addr &&
-                    instr_is_syscall(shard->prev_instr_decoded_)) {
-                    // Set this flag so that repeated syscalls are not
-                    // double reporeted as other PC discontinuity below.
-                    saw_repeated_syscall_instrs_with_same_pc = true;
-                    report_if_false(shard, false,
-                                    "Repeated syscall instrs with the same PC");
-                }
-            }
+        if (shard->prev_instr_.instr.addr != 0 /*first*/ &&
+            instr_is_syscall(cur_instr_decoded.get()) &&
+            memref.instr.addr == shard->prev_instr_.instr.addr &&
+            instr_is_syscall(shard->prev_instr_decoded_) &&
+            cur_instr_decoded != nullptr && shard->prev_instr_decoded_ != nullptr) {
```

Comment:
These `!= nullptr` checks should be moved to be before the `instr_is_syscall` checks (so that those calls are skipped for nullptrs)